src/main/scala/com/wavesplatform/matcher/model/OrderValidator.scala

```diff
@@ -133,17 +144,68 @@ object OrderValidator {
     )
   }
 
+  private[matcher] def getValidFeeAsset(order: Order, assetType: AssetType): Option[AssetId] = {
+    assetType match {
+      case AssetType.AMOUNT    => order.assetPair.amountAsset
+      case AssetType.PRICE     => order.assetPair.priceAsset
+      case AssetType.RECEIVING => order.getReceiveAssetId
+      case AssetType.SPENDING  => order.getSpendAssetId
+    }
+  }
+
+  private[matcher] def getMinValidFee(order: Order, percentSettings: PercentSettings): Long = {
+
+    lazy val receiveAmount = order.getReceiveAmount(order.amount, order.price).explicitGet()
+    lazy val spentAmount   = order.getSpendAmount(order.amount, order.price).explicitGet()
+
+    val amountFactor = percentSettings.assetType match {
+      case AssetType.AMOUNT    => order.amount
+      case AssetType.PRICE     => if (order.orderType == OrderType.BUY) spentAmount else receiveAmount
+      case AssetType.RECEIVING => receiveAmount
+      case AssetType.SPENDING  => spentAmount
+    }
+
+    multiplyLongByDouble(amountFactor, percentSettings.minFee / 100)
+  }
+
+  def validateOrderFee(order: Order, orderFeeSettings: OrderFeeSettings): ValidationResult = {
+    for {
+      _ <- (Right(order): ValidationResult)
+        .ensure(s"Matcher's fee asset in order (${order.matcherFeeAssetId}) does not meet matcher's settings requirements") { order =>
+          val isMatcherFeeAssetValid = orderFeeSettings match {
+            case _: FixedWavesSettings            => order.matcherFeeAssetId.isEmpty
+            case FixedSettings(defaultAssetId, _) => order.matcherFeeAssetId == defaultAssetId
+            case PercentSettings(assetType, _)    => order.matcherFeeAssetId == getValidFeeAsset(order, assetType)
+          }
+          order.version != 3 || isMatcherFeeAssetValid
+        }
+      _ <- (Right(order): ValidationResult)
+        .ensure(s"Matcher's fee (${order.matcherFee}) is less than minimally admissible one") { order =>
+          val isMatcherFeeValid = orderFeeSettings match {
+            case FixedWavesSettings(wavesMinFee)  => order.matcherFee >= wavesMinFee
```

Comment:
Actually, we should add to the `wavesMinFee` additional `CommonValidation.ScriptExtraFee` for each paragraph:
1. Scripted matcher's account
2. Scripted amount asset
3. Scripted price asset

Reply:
Also it looks like this validation should be in `blockchainAware` by this reason :(

Reply:
We already have such validation in `blockchainAware`